heed-traits/src/lib.rs

```diff
@@ -25,3 +26,54 @@ pub trait BytesDecode<'a> {
     /// Decode the given bytes as DItem
     fn bytes_decode(bytes: &'a [u8]) -> Result<Self::DItem, BoxedError>;
 }
+
+/// Define a custom key comparison function for a database.
+///
+/// The comparison function is called whenever it is necessary to compare a key specified
+/// by the application with a key currently stored in the database. If no comparison function
+/// is specified, and no special key flags were specified, the keys are compared lexically,
+/// with shorter keys collating before longer keys.
+pub trait Comparator {
+    /// Compares the raw bytes representation of two keys.
+    ///
+    /// # Safety
+    ///
+    /// This function must never crash, this is the reason why it takes raw bytes as parameter,
+    /// to let you define the recovery method you want in case of a decoding error.
+    fn compare(a: &[u8], b: &[u8]) -> Ordering;
+}
+
+/// Define a lexicographic comparator, which is a special case of [`Comparator`].
+///
+/// Types that implements [`LexicographicComparator`] will automatically have [`Comparator`]
+/// implemented as well, where the [`Comparator::compare`] is implemented per the definition
+/// of lexicographic ordering with [`LexicographicComparator::compare_elem`].
+///
+/// This trait is introduced to support prefix iterators, which implicit assumes that the
+/// underlying key comparator is lexicographic.
+pub trait LexicographicComparator: Comparator {
+    /// Compare a single byte; this function is used to implement [`Comparator::compare`]
+    /// by definition of lexicographic ordering.
+    ///
+    /// # Safety
+    ///
+    /// This function must never crash.
+    fn compare_elem(a: u8, b: u8) -> Ordering;
+
+    /// Update `bytes` to its immediate successor.
+    fn advance(bytes: &mut Vec<u8>);
+
+    /// Update `bytes` to its immediate predecessor.
+    fn retreat(bytes: &mut Vec<u8>);
+}
+
+impl<C: LexicographicComparator> Comparator for C {
+    fn compare(a: &[u8], b: &[u8]) -> Ordering {
+        for idx in 0..std::cmp::min(a.len(), b.len()) {
+            if a[idx] != b[idx] {
+                return a[idx].cmp(&b[idx]);
+            }
+        }
+        a.len().cmp(&b.len())
```

Comment:
Shouldn't we use [the `[u8]::cmp`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.cmp) method here?
```suggestion
        a.cmp(b)
```

Reply:
Use `Ord::Cmp` instead here, which is indeed more readable!